[QoI] Improve diagnostics of unsatisfied generic requirements

xedin opened the pull request:
Fixes a problem related to presence of InOutType in function parameters
which diagnostics related to generic parameter requirements didn't handle
correctly, and improves diagnostics for unsatisfied generic requirements
in operator applications, which we didn't attempt to diagnose at all.

Resolves: rdar://problem/33477726

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[QoI] Improve diagnostics of unsatisfied generic requirements

Fixes a problem related to presence of InOutType in function parameters
which diagnostics related to generic parameter requirements didn't handle
correctly, and improves diagnostics for unsatisfied generic requirements
in operator applications, which we didn't attempt to diagnose at all.

Resolves: rdar://problem/33477726

xedin commented:
@swift-ci please smoke test

xedin commented:
@jrose-apple @DougGregor I'm not sure if that's a good idea but I couldn't figure out how to word the diagnostic so it mentions *what* exactly went wrong and *how*. Instead I made two diagnostics - one which says that we can't apply_operator/call_function with given parameters and another one clarifies why - because one of the requirements haven't been met.

DougGregor commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5980,18 +5987,69 @@ bool FailureDiagnosis::diagnoseArgumentGenericRequirements(
+      switch (req.getKind()) {
+      case RequirementKind::Conformance:
+      case RequirementKind::Layout:
+        return;
```
I bet you could customize the `Conformance` case here to improve diagnostics when one of the type arguments (or associated types thereof) fails to meet some conformance requirement.

DougGregor commented on lib/Sema/CSDiag.cpp:
```diff
@@ -5980,18 +5987,69 @@ bool FailureDiagnosis::diagnoseArgumentGenericRequirements(
+        TC.diagnose(ArgExpr->getLoc(), diag::cannot_apply_unop_to_arg,
+                    overloadName, CS.getType(ArgExpr));
+      } else {
+        bool isInitializer = isa<TypeExpr>(FnExpr);
```
How about asking whether `isa<ConstructorDecl>(AFD)`?

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -1252,6 +1256,8 @@ RequirementCheckResult TypeChecker::checkGenericArguments(
       case RequirementCheckResult::Failure:
       case RequirementCheckResult::SubstitutionFailure:
+        if (listener)
+          listener->unsatisfiedRequirement(rawReq, firstType, secondType);
```
I don't think we only want to call `unsatisfiedRequirement` for `UnsatisfiedDependency`.

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -1274,37 +1280,36 @@ RequirementCheckResult TypeChecker::checkGenericArguments(
       continue;
+
+    if (listener)
+      listener->unsatisfiedRequirement(rawReq, firstType, secondType);
```
The implied contract for `unsatisfiedRequirement` is a bit odd... it might diagnose the problem, or it might not, and `checkGenericArguments` will (separately) decide based on the validity of `loc` whether it should diagnose. Could we have `unsatisfiedRequirement` return `true` to indicate when it diagnosed the requirement, in which case `checkGenericArguments` won't emit a diagnostic? Perhaps then we should call it `diagnoseUnsatisfiedRequirement`.

DougGregor commented on test/Constraints/generics.swift:
```diff
@@ -251,7 +251,10 @@ struct V27515965 {
-  return V27515965(x) // expected-error {{generic parameter 'T' could not be inferred}}
+  return V27515965(x)
+  // expected-error@-1 {{cannot invoke initializer for type 'init(_:)' with an argument list of type '(S27515965)'}}
+  // expected-error@-2 {{'<T where T : P27515965, T.R == Float> (V27515965.Type) -> (T) -> V27515965' requires the types 'Any' and 'Float' be equivalent}}
```
Shouldn't this second error actually be a note? It's clarifying the previous error. Also, we shouldn't be printing the generic signature here. We'd like to point at the candidate and say something that describes the problem and how we got there within just one note (so we don't have the third diagnostic here):

    candidate requires that the types 'Any' and 'Float' be equivalent (i.e., 'T.R' == 'Float' with T == S27515965)

that approach scales to multiple candidates.

DougGregor reviewed (changes requested):
Definitely progress! Let's tune these to make them great.

xedin commented:
@DougGregor Addressed most of the comments, expect for conformance one, because that would require changes in the `conformsToProtocol` and we already produce pretty good message, will change it once I get into `CalleeCandidateInfo` refactoring. WDYT?

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -1250,6 +1254,9 @@ RequirementCheckResult TypeChecker::checkGenericArguments(
       case RequirementCheckResult::UnsatisfiedDependency:
+          if (listener)
+            listener->diagnoseUnsatisfiedRequirement(rawReq, firstType, secondType);
+          LLVM_FALLTHROUGH;
```
This `diagnoseUnsatisfiedRequirement` isn't in the right place; it's `Failure` and `SubstitutionFailure` that would be diagnosed. Although at this point, it's dead code because you don't have a listener anywhere that diagnoses conformance failures, so I'd be okay with deleting this bit and adding a FIXME.

xedin replied:
Done!

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -1274,37 +1281,37 @@ RequirementCheckResult TypeChecker::checkGenericArguments(
+      diagnose(noteLoc, diagnosticNote, rawFirstType, rawSecondType,
+               genericSig->gatherGenericParamBindingsText(
+                   {rawFirstType, rawSecondType}, substitutions));
     }
```
Very nice!

DougGregor commented on test/Generics/deduction.swift:
```diff
@@ -216,11 +216,12 @@ func callRangeOfIsBefore(_ ia: [Int], da: [Double]) {
 
 func testEqualIterElementTypes<A: IteratorProtocol, B: IteratorProtocol>(_ a: A, _ b: B) where A.Element == B.Element {}
-// expected-note@-1 {{requirement specified as 'A.Element' == 'B.Element' [with A = IndexingIterator<[Int]>, B = IndexingIterator<[Double]>]}}
+// expected-note@-1 {{candidate requires that the types 'Int' and 'Double' be equivalent (requirement specified as 'A.Element' == 'B.Element' [with A = IndexingIterator<[Int]>, B = IndexingIterator<[Double]>])}}
```
These are fantastic, thank you!

xedin commented:
@swift-ci please smoke test and merge

xedin commented:
@swift-ci please smoke test and merge
